### PR TITLE
8295166: IGV: dump graph at more locations

### PR DESCRIPTION
**Pull request by dlunde**
This changeset
1. adds a number of new graph dumps for IdealGraphVisualizer (IGV):
   - Before conditional constant propagation
   - After register allocation
   - After block ordering
   - After peephole optimization
   - After post-allocation expansion
   - Before and after
     - loop predication
     - loop peeling
     - pre/main/post loops
     - loop unrolling
     - range check elimination
     - loop unswitching
     - partial peeling
     - split if
     - superword
2. adds support for enumeration of repeated IGV graph dumps.
3. adjusts IGV print levels to encompass the new graph dumps. The old levels 4 and 5 are now levels 5 and 6. The new level 4 is for loop optimization dumps.

Example phase list screenshots in IGV (first at level 6, second at level 4)
Screenshot from 2023-12-04 13-55-38 Screenshot from 2023-12-04 13-56-29

Some notes:
- While discussing the above changes, a separate question was brought up by @chhagedorn:
  > On a separate note, I'm wondering how useful it is to always dump all JFR events when calling print_method(). Should this be revisited again in general?
- The new IGV graph dump enumeration enables a number of cleanups. There is now another RFE for IGV cleanup: [JDK-8319599](https://bugs.openjdk.org/browse/JDK-8319599).

#### Platforms: windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64
- tier1, tier2, tier3, tier4, tier5.
- Check that optimized builds (`--with-debug-level optimized`) still work.

#### Platforms: linux-x64
- Tested that thousands of graphs are correctly opened and visualized with IGV.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295166](https://bugs.openjdk.org/browse/JDK-8295166): IGV: dump graph at more locations (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [07aac1c5](https://git.openjdk.org/jdk/pull/16120/files/07aac1c538fe8673dccfd4a99d32f5e6a1b9150b)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [6f872e5d](https://git.openjdk.org/jdk/pull/16120/files/6f872e5dfdd6b6189d2a25cd407ad2645db3b511)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16120/head:pull/16120` \
`$ git checkout pull/16120`

Update a local copy of the PR: \
`$ git checkout pull/16120` \
`$ git pull https://git.openjdk.org/jdk.git pull/16120/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16120`

View PR using the GUI difftool: \
`$ git pr show -t 16120`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16120.diff">https://git.openjdk.org/jdk/pull/16120.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16120#issuecomment-1840676073)

### CODE REVIEW

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -622,6 +622,7 @@ Compile::Compile( ciEnv* ci_env, ciMethod* target, int osr_bci,
                   _has_reserved_stack_access(target->has_reserved_stack_access()),
 #ifndef PRODUCT
                   _igv_idx(0),
+                  _igv_phase_iter(),
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:625]: This is value initialization which guarantees proper zeroing, right? For other arrays, for example `Compile::_trap_hist`, we use explicit `Copy::zero_to_bytes` but I think your variant is fine.

File: src/hotspot/share/opto/compile.cpp [hunk 2]
```diff
@@ -5094,6 +5107,15 @@ void Compile::print_method(CompilerPhaseType cpt, int level, Node* n) {
   C->_latest_stage_start_counter.stamp();
 }
 
+void Compile::print_method_iter(CompilerPhaseType cpt, int level, Node* n) {
+#ifndef PRODUCT
+  int iter = ++_igv_phase_iter[cpt];
+  print_method(cpt, level, n, iter);
+#else
+  print_method(cpt, level, n);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:5115]: This is dead code because all calls are guarded by `NOT_PRODUCT`, right?

File: src/utils/IdealGraphVisualizer/.gitignore [hunk 3]
```diff
@@ -1,2 +1,6 @@
 /*/target
-.igv.log
\ No newline at end of file
+.igv.log
+/all-*.dat
+/lastModified/
+/localeVariants
+/package-attrs.dat
```
Comment (TobiHartmann) [inline src/utils/IdealGraphVisualizer/.gitignore:6]: Is that really needed? Just wondering why these files haven't been added before.

**Review by TobiHartmann — COMMENTED** [review]
Nice enhancement! Looks good to me overall. Some comments below.

Please add an IGV screenshot of how the new phases look in the phase list.

Could you summarize which of the ideas / proposals from the RFE are not covered? We should file a follow-up RFE for them.

> Many of the proposed dump locations are educated guesses. Should we adjust any of them?

They look good to me but let's see what others think.

> Are the proposed levels (for PrintIdealGraphLevel) reasonable or should we adjust them? I put the loop optimization dumps at level 4 and adjusted IdealGraphVisualizer/README.md accordingly.

I think that's reasonable.

> I put most new calls to print_method/print_method_iter within a NOT_PRODUCT. Is this OK?

Existing calls to `print_method` are not guarded because the method also commits s JFR event and updates `Compile::_latest_stage_start_counter`, so I think your new code should behave similar.

And please make sure that you verify that the 'optimized' build (`--with-debug-level optimized`) still works. It's a level between fastdebug and release where both `#ifdef ASSERT` and `#ifdef PRODUCT` are false.

**Comment by dlunde** [comment]
Thanks for the review Tobias!

> Please add an IGV screenshot of how the new phases look in the phase list.

I've attached two sample screenshots of the phase list (from running one of the tests in `hotspot/jtreg/compiler/loopopts/PartialPeelingUnswitch.java`). Note the enumeration of the loop optimizations and the attached target loop node indices.

Screenshot from 2023-10-13 12-59-22
Screenshot from 2023-10-13 12-59-53


> Could you summarize which of the ideas / proposals from the RFE are not covered? We should file a follow-up RFE for them.

I believe I have covered everything in the RFE.

> Existing calls to print_method are not guarded because the method also commits a JFR event and updates Compile::_latest_stage_start_counter, so I think your new code should behave similar.

Ah, I see. Note, however, that the call to `print_method` for `PHASE_AFTER_ITER_GVN_STEP` (at level 4) is guarded. Also, the bytecode printing (level 5) is guarded (although it does not use `Compile::print_method`).

> And please make sure that you verify that the 'optimized' build (--with-debug-level optimized) still works. It's a level between fastdebug and release where both #ifdef ASSERT and #ifdef PRODUCT are false.

I'll make sure to include this when testing.

File: src/utils/IdealGraphVisualizer/.gitignore (hunk 3 above)
Comment (dlunde) [reply src/utils/IdealGraphVisualizer/.gitignore:6]: Not sure, IGV generates them at first run on my system (after a `git clean` and clean IGV build).

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp:5115]: Not quite, there is one unguarded call for `PHASE_PHASEIDEALLOOP_ITERATIONS` at level 2. This is an existing phase that I converted from `print_method` to `print_method_iter` (as suggested by @chhagedorn in the issue corresponding to this PR).

File: src/hotspot/share/opto/compile.cpp (hunk 1 above)
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp:625]: Yes, correct. I'll switch to the `Copy::zero_to_bytes` initialization, better to be consistent.

**Review by robcasloz — COMMENTED** [review]
Hi Daniel, thanks for working on this! The code changes themselves look good, I have a few comments/suggestions:
- It might make sense to create a new print level between current 3 and 4 including the new loop transformation dumps but not the individual IGVN step dumps.
- I see the value in numbering the `PHASE_PHASEIDEALLOOP_ITERATIONS` dumps, but I am not sure numbering the new loop transformations is worth the additional complexity (`_igv_phase_iter` array, etc.). Limiting numbering to `PHASE_PHASEIDEALLOOP_ITERATIONS` dumps would not require any additional state in `Compile`.
- In my opinion, it would be clearer to only dump loop transformations if they actually take place. I find it a bit confusing to see e.g. `Before/After superword ...` graph dumps when vectorization has actually failed and the graph has not changed at all. Dumping only after effective transformations would also match better the output of `TraceLoopOpts`. Enforcing this invariant would require getting rid of the `Before...`  dumps though, but this is acceptable (and perhaps even preferable) in my opinion.

File: src/hotspot/share/opto/loopopts.cpp [hunk 4]
```diff
@@ -3895,6 +3898,9 @@ bool PhaseIdealLoop::partial_peel( IdealLoopTree *loop, Node_List &old_new ) {
     }
   }
 #endif
+
+  NOT_PRODUCT(C->print_method_iter(PHASE_PARTIAL_PEEL, 4, head);)
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:3902]: Here you can also dump the new head `new_head_clone` after partial peeling.

**Review by chhagedorn — COMMENTED** [review]
Thanks for working on this! This is a good addition and helps to better debug with IGV. I left a few comments with some suggestion and improvement ideas.

File: src/hotspot/share/opto/phasetype.hpp [hunk 5]
```diff
@@ -26,51 +26,74 @@
...
-  flags(BEFORE_MATCHING,              "Before matching") \
-  flags(MATCHING,                     "After matching") \
-  flags(GLOBAL_CODE_MOTION,           "Global code motion") \
-  flags(MACH_ANALYSIS,                "After mach analysis") \
-  flags(FINAL_CODE,                   "Final Code") \
-  flags(END,                          "End") \
-  flags(FAILURE,                      "Failure") \
-  flags(ALL,                          "All") \
-  flags(DEBUG,                        "Debug")
+  flags(BEFORE_STRINGOPTS,              "Before StringOpts") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:29]: I believe I've kept it consistent for my own additions, but the older phase names are sometimes inconsistent in this regard (including `STRINGOPTS`). Should I rename other phases to improve consistency? This changeset will touch many more files then, but perhaps that's OK.

[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:29]: My mistake, it was actually called `BEFORE_STRINGOPTS` and `AFTER_STRINGOPTS`. But, the old phases `MATCHING` and `MACH_ANALYSIS` do contain "After" in the string but not in the name.

File: src/hotspot/share/opto/compile.cpp [hunk 6]
```diff
@@ -5079,6 +5092,9 @@ void Compile::print_method(CompilerPhaseType cpt, int level, Node* n) {
   ResourceMark rm;
   stringStream ss;
   ss.print_raw(CompilerPhaseTypeHelper::to_description(cpt));
+  if (iter > 0) {
+    ss.print(" %d", iter);
+  }
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp:5097]: I agree with just having a `print_method`. I initially hesitated to modify `print_method` in the way you suggest, as it would then add iteration numbering to _all_ phases with no exceptions. But, it seems this is a feature we want then?

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp:5118]: I have now incorporated the functionality of `print_method_iter` directly into `print_method`, and also removed all the `NOT_PRODUCT` wrappers.

I'm resolving this thread now, should we move the discussion regarding JFR event dumping somewhere else?

File: src/utils/IdealGraphVisualizer/.gitignore (hunk 3 above)
Comment (dlunde) [reply src/utils/IdealGraphVisualizer/.gitignore:6]: I'll remove this from the PR and make a separate JBS issue.

File: src/hotspot/share/opto/phasetype.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:29]: I'll resolve this by using `BEFORE_` and `AFTER_` for all new phases that I'm adding, but leaving existing phase names intact. We can address inconsistencies in another RFE.

File: src/hotspot/share/opto/phasetype.hpp [hunk 7]
```diff
@@ -26,51 +26,74 @@
...
+  flags(AFTER_CLOOPS,                   "After CountedLoop") \
+  flags(PHASEIDEAL_BEFORE_EA,           "PhaseIdealLoop before EA") \
+  flags(AFTER_EA,                       "After Escape Analysis") \
+  flags(ITER_GVN_AFTER_EA,              "Iter GVN after EA") \
+  flags(ITER_GVN_AFTER_ELIMINATION,     "Iter GVN after eliminating allocations and locks") \
+  flags(PHASEIDEALLOOP1,                "PhaseIdealLoop 1") \
+  flags(PHASEIDEALLOOP2,                "PhaseIdealLoop 2") \
+  flags(PHASEIDEALLOOP3,                "PhaseIdealLoop 3") \
+  flags(BEFORE_CCP1,                    "Before PhaseCCP 1") \
+  flags(CCP1,                           "PhaseCCP 1") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:78]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/compile.cpp]: Maybe we can mention it as a side node in the PR description that it is something we should think about at some point. But it should not block this PR.

File: src/hotspot/share/opto/compile.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (chhagedorn) [reply src/hotspot/share/opto/compile.cpp]: I think it could be useful for any repeated phase, if others also agree with that. And we would still keep the same name for phases that are only printed once and only add a number for the repeated dumps.

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp]: Sure, I'll add it to the final (non-draft) PR description.

File: src/hotspot/share/opto/compile.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/compile.cpp]: I've changed it for now. If anyone does not agree, please let us know.

File: src/hotspot/share/opto/phasetype.hpp [hunk 8]
```diff
@@ -26,51 +26,74 @@
...
+  flags(SUPERWORD,                      "After superword")\
+  flags(BEFORE_CLOOPS,                  "Before CountedLoop") \
+  flags(AFTER_CLOOPS,                   "After CountedLoop") \
+  flags(PHASEIDEAL_BEFORE_EA,           "PhaseIdealLoop before EA") \
+  flags(AFTER_EA,                       "After Escape Analysis") \
+  flags(ITER_GVN_AFTER_EA,              "Iter GVN after EA") \
+  flags(ITER_GVN_AFTER_ELIMINATION,     "Iter GVN after eliminating allocations and locks") \
+  flags(PHASEIDEALLOOP1,                "PhaseIdealLoop 1") \
+  flags(PHASEIDEALLOOP2,                "PhaseIdealLoop 2") \
+  flags(PHASEIDEALLOOP3,                "PhaseIdealLoop 3") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:76]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/phasetype.hpp [hunk 9]
```diff
@@ -26,51 +26,74 @@
...
-  flags(END,                          "End") \
-  flags(FAILURE,                      "Failure") \
-  flags(ALL,                          "All") \
-  flags(DEBUG,                        "Debug")
+  flags(BEFORE_STRINGOPTS,              "Before StringOpts") \
+  flags(AFTER_STRINGOPTS,               "After StringOpts") \
+  flags(BEFORE_REMOVEUSELESS,           "Before RemoveUseless") \
+  flags(AFTER_PARSING,                  "After Parsing") \
+  flags(BEFORE_ITER_GVN,                "Before Iter GVN") \
+  flags(ITER_GVN1,                      "Iter GVN 1") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:34]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/phasetype.hpp [hunk 10]
```diff
@@ -26,51 +26,74 @@
...
+  flags(AFTER_ITER_GVN_STEP,            "After Iter GVN Step") \
+  flags(AFTER_ITER_GVN,                 "After Iter GVN") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:36]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/phasetype.hpp [hunk 11]
```diff
@@ -26,51 +26,74 @@
...
+  flags(RANGE_CHECK_ELIMINATION,        "After range check elimination") \
+  flags(BEFORE_PRE_POST_LOOPS,          "Before pre/post loops") \
+  flags(PRE_POST_LOOPS,                 "After pre/post loops") \
+  flags(BEFORE_SUPERWORD,               "Before superword")\
+  flags(SUPERWORD,                      "After superword")\
+  flags(BEFORE_CLOOPS,                  "Before CountedLoop") \
+  flags(AFTER_CLOOPS,                   "After CountedLoop") \
+  flags(PHASEIDEAL_BEFORE_EA,           "PhaseIdealLoop before EA") \
+  flags(AFTER_EA,                       "After Escape Analysis") \
+  flags(ITER_GVN_AFTER_EA,              "Iter GVN after EA") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:72]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/phasetype.hpp [hunk 12]
```diff
@@ -26,51 +26,74 @@
...
+  flags(ITER_GVN_AFTER_ELIMINATION,     "Iter GVN after eliminating allocations and locks") \
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp:73]: Will be addressed in a separate RFE.

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 13]
```diff
@@ -1387,6 +1387,8 @@ bool PhaseIdealLoop::loop_predication_impl(IdealLoopTree* loop) {
   ConNode* zero = _igvn.intcon(0);
   set_ctrl(zero, C->root());
 
+  NOT_PRODUCT(C->print_method_iter(PHASE_BEFORE_LOOP_PREDICATION, 4, head);)
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/loopPredicate.cpp]: Updated now

File: src/hotspot/share/opto/loopopts.cpp (hunk 4 above)
Comment (dlunde) [reply src/hotspot/share/opto/loopopts.cpp]: Thanks, updated.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 14]
```diff
@@ -2374,6 +2386,8 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
     }
   }
 #endif
+
+  NOT_PRODUCT(C->print_method_iter(PHASE_UNROLL_LOOP, 4, loop_head);)
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/loopTransform.cpp]: Updated

File: src/hotspot/share/opto/loopTransform.cpp [hunk 15]
```diff
@@ -2855,6 +2869,8 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
   assert(RangeCheckElimination, "");
   CountedLoopNode *cl = loop->_head->as_CountedLoop();
 
+  NOT_PRODUCT(C->print_method_iter(PHASE_BEFORE_RANGE_CHECK_ELIMINATION, 4, cl);)
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/hotspot/share/opto/loopTransform.cpp]: Updated now

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 16]
```diff
@@ -1270,6 +1273,7 @@ bool PhaseIdealLoop::loop_predication_impl_helper(IdealLoopTree* loop, IfProjNod
     new_predicate_proj = add_template_assertion_predicate(iff, loop, if_success_proj, parse_predicate_proj, upper_bound_proj, scale,
                                                           offset, init, limit, stride, rng, overflow, reason);
 
+    C->print_method(PHASE_AFTER_LOOP_PREDICATION_RC, 4, new_predicate_proj->in(0));
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopPredicate.cpp:1276]: I thought about this here again. I propose to merge `PHASE_AFTER_LOOP_PREDICATION_RC` and `PHASE_AFTER_LOOP_PREDICATION_IC` to a single `PHASE_AFTER_LOOP_PREDICATION` phase and dump it after `dominated_by()` on L1289. which kills the hoisted check. I think that's more intuitive. Otherwise, the old and the new `If` still share the same `BoolNode` in the dump. You can use `new_predicate_proj->in(0)` which is the same as `new_predicate_iff` for the invariant check.

File: src/hotspot/share/opto/phasetype.hpp [hunk 17]
```diff
@@ -26,51 +26,77 @@
...
+  flags(SCALARIZE_VBOX,                 "Scalarize VectorBox") \
+  flags(INLINE_VECTOR_REBOX,            "Inline Vector Rebox Calls") \
+  flags(EXPAND_VBOX,                    "Expand VectorBox") \
+  flags(ELIMINATE_VBOX_ALLOC,           "Eliminate VectorBoxAllocate") \
+  flags(ITER_GVN_BEFORE_EA,             "Iter GVN before EA") \
+  flags(ITER_GVN_AFTER_VECTOR,          "Iter GVN after vector box elimination") \
+  flags(BEFORE_BEAUTIFY_LOOPS,          "Before beautify loops") \
+  flags(AFTER_BEAUTIFY_LOOPS,           "After beautify loops") \
+  flags(BEFORE_LOOP_UNROLLING,          "Before loop unrolling") \
+  flags(AFTER_LOOP_UNROLLING,           "After loop unrolling") \
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:51]: Nits: I suggest to use upper case letters for nouns in the new phase name strings to follow the convention of the other existing phases.

Comment (dlunde) [reply src/hotspot/share/opto/phasetype.hpp]: I switched the new phases to title case. The existing phases are not quite consistent either, so I suggest that we change all phase descriptions to title case (I've added an item for this to the IGV cleanup RFE)

File: src/hotspot/share/opto/loopPredicate.cpp (hunk 16 above)
Comment (dlunde) [reply src/hotspot/share/opto/loopPredicate.cpp:1276]: Good, now fixed. Just to be clear, I merged the two `AFTER` phases but did not touch the `BEFORE` phases.

File: test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java [hunk 18]
```diff
@@ -60,6 +60,26 @@ public enum CompilePhase {
     ITER_GVN_AFTER_VECTOR("Iter GVN after vector box elimination"),
     BEFORE_BEAUTIFY_LOOPS("Before beautify loops"),
     AFTER_BEAUTIFY_LOOPS("After beautify loops"),
+    BEFORE_LOOP_UNROLLING("Before Loop Unrolling"),
+    AFTER_LOOP_UNROLLING("After Loop Unrolling"),
+    BEFORE_SPLIT_IF("Before Split If"),
+    AFTER_SPLIT_IF("After Split If"),
+    BEFORE_LOOP_PREDICATION_IC("Before Loop Predication IC"),
+    BEFORE_LOOP_PREDICATION_RC("Before Loop Predication RC"),
+    AFTER_LOOP_PREDICATION("After Loop Predication"),
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java:69]: Would it be possible to define `AFTER_LOOP_PREDICATION_IC` and `AFTER_LOOP_PREDICATION_RC` separately, to match the corresponding `BEFORE_*` phases? This would make it easier to understand what is the scope of each transformation in the IGV outline and also play better with the enumeration of repeated IGV graph dumps introduced in this changeset. To be more specific, what I would expect to see here:
igv-pred
is:
```
15. Before Loop Predication IC: 90 If
16. After Loop Predication IC: 163 If
17. Before Loop Predication RC: 106 RangeCheck
18. After Loop Predication RC: 196 RangeCheck
```

File: test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java [hunk 19]
```diff
@@ -81,6 +102,10 @@ public enum CompilePhase {
     MATCHING("After matching", RegexType.MACH),
     MACH_ANALYSIS("After mach analysis", RegexType.MACH),
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java:103]: Could you move this line to after `POSTALLOC_EXPAND("Post-Allocation Expand", RegexType.MACH)`, for consistency with the order in which phases are defined in `phasetype.hpp`?

File: src/hotspot/share/opto/compile.cpp [hunk 20]
```diff
@@ -2989,19 +2996,25 @@ void Compile::Code_Gen() {
...
+
+    print_method(PHASE_BLOCK_ORDERING, 3);
   }
 
   // Apply peephole optimizations
   if( OptoPeephole ) {
     TracePhase tp("peephole", &timers[_t_peephole]);
     PhasePeephole peep( _regalloc, cfg);
     peep.do_transform();
+
```
Comment (robcasloz) [inline src/hotspot/share/opto/compile.cpp:3008]: Nit: no need for extra line here.

File: test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java (hunk 18 above)
Comment (dlunde) [reply test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java:69]: Sounds good to me, this is how we implemented it originally. I see @chhagedorn reacted with a thumbs up, so I'll go ahead with the change!

File: src/utils/IdealGraphVisualizer/README.md [hunk 21]
```diff
@@ -30,8 +30,9 @@ Ideal graphs are dumped at the following points:
   compilations, if available)
 * `N=2`: additionally, after every major phase (including loop opts)
```
[the remark this answers is not in the document]
Comment (dlunde) [reply src/utils/IdealGraphVisualizer/README.md]: Fixed, thanks

File: test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java (hunk 19 above)
Comment (dlunde) [reply test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java]: Yes, fixed now. In general, I suspect there are other inconsistencies between `CompilePhase.java` and `phasetype.hpp`. I'll add it as an item to the cleanup RFE.

File: src/hotspot/share/opto/loopopts.cpp [hunk 22]
```diff
@@ -1446,7 +1446,12 @@ void PhaseIdealLoop::split_if_with_blocks_post(Node *n) {
     }
 
     // Now split the IF
+    C->print_method(PHASE_BEFORE_SPLIT_IF, 4, iff);
+    if ((PrintOpto && VerifyLoopOptimizations) || TraceLoopOpts) {
+      tty->print_cr("Split-if");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:1451]:
```suggestion
      tty->print_cr("Split-If");
```

File: src/hotspot/share/opto/phasetype.hpp [hunk 23]
```diff
@@ -28,51 +28,77 @@
...
+  flags(EXPAND_VBOX,                    "Expand VectorBox") \
+  flags(ELIMINATE_VBOX_ALLOC,           "Eliminate VectorBoxAllocate") \
+  flags(ITER_GVN_BEFORE_EA,             "Iter GVN before EA") \
+  flags(ITER_GVN_AFTER_VECTOR,          "Iter GVN after vector box elimination") \
+  flags(BEFORE_BEAUTIFY_LOOPS,          "Before beautify loops") \
+  flags(AFTER_BEAUTIFY_LOOPS,           "After beautify loops") \
+  flags(BEFORE_LOOP_UNROLLING,          "Before Loop Unrolling") \
+  flags(AFTER_LOOP_UNROLLING,           "After Loop Unrolling") \
+  flags(BEFORE_SPLIT_IF,                "Before Split If") \
+  flags(AFTER_SPLIT_IF,                 "After Split If") \
```
Comment (chhagedorn) [inline src/hotspot/share/opto/phasetype.hpp:55]: I suggest to use a `-`:
```suggestion
  flags(BEFORE_SPLIT_IF,                "Before Split-If") \
  flags(AFTER_SPLIT_IF,                 "After Split-If") \
```

File: test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java [hunk 24]
```diff
@@ -60,6 +60,27 @@ public enum CompilePhase {
     ITER_GVN_AFTER_VECTOR("Iter GVN after vector box elimination"),
     BEFORE_BEAUTIFY_LOOPS("Before beautify loops"),
     AFTER_BEAUTIFY_LOOPS("After beautify loops"),
+    BEFORE_LOOP_UNROLLING("Before Loop Unrolling"),
+    AFTER_LOOP_UNROLLING("After Loop Unrolling"),
+    BEFORE_SPLIT_IF("Before Split If"),
+    AFTER_SPLIT_IF("After Split If"),
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java:66]:
```suggestion
    BEFORE_SPLIT_IF("Before Split-If"),
    AFTER_SPLIT_IF("After Split-If"),
```

**Review by chhagedorn — APPROVED** [review]
Otherwise, it looks good to me! Thanks for addressing all the comments and suggestions.

**Review by chhagedorn — APPROVED** [review]
